### PR TITLE
Add cpp performance test server / client

### PR DESCRIPTION
**Pull request by achim-k**
### Public-Facing Changes

Add cpp performance test server / client

### Description
Adds a server & client for testing websocket throughput. The server advertises various channels (low-bandwidth to high-bandwidth) which can be subscribed by the test client or https://foxglove.github.io/ws-protocol/ for measuring websocket throughput.

```
$ rosrun foxglove_bridge perf_test_server --help
Usage: /ros_ws/install_ros1/lib/foxglove_bridge/perf_test_server [options]
Options:
  --address arg (=0.0.0.0)            Address to bind to
  --port arg (=8765)                  Port to listen to
  --send-buffer-limit arg (=20000000) Send buffer limit in bytes
  --channel-config arg                Path to JSON file with custom channels in
                                      the format {topic: string, frequency: 
                                      number, sizeInBytes:number}[].
  --compression                       Enable per-message deflate compression
  --certfile arg                      Path to the certificate to use for TLS
  --keyfile arg                       Path to the private key to use for TLS
  --help                              Produce help message
```

```
$ rosrun foxglove_bridge perf_test_client 
Connected to ws://localhost:8765
^C
|             topic | Msgs rcvd |   bytes rcvd | bandwidth (b/s) |
|-------------------|-----------|--------------|-----------------|
| 10 bytes @ 100 Hz |      1043 |        23989 |        2.27e+03 |
| 10 bytes @ 500 Hz |      4858 |       111734 |        1.05e+04 |
|    10 KB @ 100 Hz |      1041 |     10423533 |        9.84e+05 |
|    10 KB @ 500 Hz |      4838 |     48442894 |        4.57e+06 |
|    100 KB @ 10 Hz |       106 |     10601378 |        1.01e+06 |
|    100 KB @ 50 Hz |       524 |     52406812 |        4.95e+06 |
|      1 MB @ 10 Hz |       106 |    106001378 |        1.00e+07 |
|      1 MB @ 50 Hz |       516 |    516006708 |        4.88e+07 |
|      5 MB @ 10 Hz |       105 |    525001365 |        5.02e+07 |
|      5 MB @ 20 Hz |       209 |   1045002717 |        9.93e+07 |
|     10 MB @ 10 Hz |       105 |   1050001365 |        1.01e+08 |
|     10 MB @ 20 Hz |       210 |   2100002730 |        2.00e+08 |
|-------------------|-----------|--------------|-----------------|
|             TOTAL |     13661 |   5464026603 |        5.20e+08 |
Connection closed
```

### CODE REVIEW

File: cpp/examples/src/example_client_perf_test.cpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  std::unordered_map<foxglove::SubscriptionId, ChannelStats> channelStatsBySubId;
+  std::unordered_map<foxglove::ChannelId, foxglove::SubscriptionId> subIdByChannelId;
+  foxglove::SubscriptionId nextSubId = 0;
+  foxglove::Client<foxglove::WebSocketNoTls> client;
+  size_t maxTopicCharLength = 0;
+
+  client.setBinaryMessageHandler([&](const uint8_t* data, size_t dataLength) {
+    if (static_cast<foxglove::BinaryOpcode>(data[0]) != foxglove::BinaryOpcode::MESSAGE_DATA) {
+      return;
+    }
```
Comment (jtbandes) [inline cpp/examples/src/example_client_perf_test.cpp:40]: Drive-by API review comment: this generic `set(Text|Binary)MessageHandler` API for the Client is not great – custom client code should not be responsible for parsing the `op`/`BinaryOpcode` itself. The point of providing a library for people to use is that it should take care of these protocol details itself so the user doesn't have to.

File: cpp/examples/src/example_client_perf_test.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        .count();
+    const double bandwidth = static_cast<double>(stats.bytesReceived) / (duration / 1000.0);
+
+    std::cout << "| " << std::setw(maxTopicCharLength) << stats.topic << " | " << std::setw(9)
+              << stats.msgsReceived << " | " << std::setw(12) << stats.bytesReceived << " | "
+              << std::setw(15) << std::setprecision(2) << std::scientific << bandwidth << " |"
+              << std::endl;
+    totalBytesRcvd += stats.bytesReceived;
+    totalMsgsRcvd += stats.msgsReceived;
+    if (std::isnormal(bandwidth)) {
```
Comment (jtbandes) [inline cpp/examples/src/example_client_perf_test.cpp:118]: why `isnormal` rather than `isfinite`?

File: cpp/examples/src/example_server_perf_test.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                     "Address to bind to")("port", po::value<int>(&port)->default_value(8765),
+                                           "Port to listen to")(
+    "send-buffer-limit", po::value<int>(&sendBufferLimit)->default_value(2e7),
+    "Send buffer limit in bytes")("channel-config", po::value<std::string>(&channelConfigFile),
+                                  "Path to JSON file with custom channels in the format "
+                                  "{topic: string, frequency: number, sizeInBytes:number}[].")(
+    "compression", "Enable per-message deflate compression")(
+    "certfile", po::value<std::string>(&certfile), "Path to the certificate to use for TLS")(
+    "keyfile", po::value<std::string>(&keyfile), "Path to the private key to use for TLS")(
+    "help", "Produce help message");
```
Comment (jtbandes) [inline cpp/examples/src/example_server_perf_test.cpp:115]: is this what clang-format does with boost program options? Pretty abysmal formatting ☹️ 

Could be worth wrapping in `// clang-format off ... // clang-format on`

Another idea is I think I might've worked around this in the past by replacing `desc.add_options()(...)(...)` with `auto&& add = desc.add_options(); add(...); add(...);`

File: cpp/foxglove-websocket/include/foxglove/websocket/websocket_client.hpp [hunk 4]
```diff
@@ -122,7 +122,8 @@ class Client : public ClientInterface {
       return;  // Already disconnected
     }
 
-    _endpoint.close(_con, websocketpp::close::status::going_away, "");
+    websocketpp::lib::error_code ec;
```
Comment (jtbandes) [inline cpp/foxglove-websocket/include/foxglove/websocket/websocket_client.hpp:125]: what's the point of this variable? does it prevent throwing an exception? Might be nice to add a comment to explain this

**Review by jtbandes — APPROVED** [review]

File: cpp/examples/src/example_client_perf_test.cpp (hunk 1 above)
Comment (achim-k) [reply cpp/examples/src/example_client_perf_test.cpp:40]: The websocket client was primarily added for use in smoke tests, so the API is rather low level. But in general I agree with you.

File: cpp/examples/src/example_client_perf_test.cpp (hunk 2 above)
Comment (achim-k) [reply cpp/examples/src/example_client_perf_test.cpp:118]: that's better, thanks!

File: cpp/foxglove-websocket/include/foxglove/websocket/websocket_client.hpp (hunk 4 above)
Comment (achim-k) [reply cpp/foxglove-websocket/include/foxglove/websocket/websocket_client.hpp:125]: Yeah, it prevented throwing an exception when calling `close` multiple times. Will add a comment

**Review by jtbandes — APPROVED** [review]
